lib/async_hooks.js

```diff
@@ -197,6 +199,141 @@ class AsyncResource {
+const contexts = [];
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // Save new resource.
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      for (let i = 0; i < contexts.length; ++i) {
+        contexts[i]._propagate(resource, triggerResource);
+      }
+    }
+  },
+  destroy(asyncId) {
+    const resource = getResource(asyncId);
+    if (resource) {
+      for (let i = 0; i < contexts.length; ++i) {
+        delete resource[contexts[i].kResourceStore];
+      }
+      idResourcePairing.delete(asyncId);
+    }
+  }
+});
+
+class AsyncContext {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (store) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enterSync() {
+    asyncContextHook.enable();
+    if (!contexts.includes(this)) {
+      contexts.push(this);
+    }
+    let resource = getCurrentResource();
+    if (!resource) {
+      // There is no resource, we need to create a fake one
+      resource = {};
+      idResourcePairing.set(executionAsyncId(), new WeakReference(resource));
+      process.nextTick(() => {
+        resource;
+      });
+      // This prevents the fake resource from being garbage collected too soon
+    }
+    const store = new Map();
+    resource[this.kResourceStore] = store;
+    return store;
+  }
+
+  _exitSync() {
+    const resource = getCurrentResource();
+    if (resource) {
+      delete resource[this.kResourceStore];
+    }
+  }
+
+  runAndReturn(callback, ...args) {
+    this._enterSync();
+    try {
+      return callback(...args);
+    } finally {
+      this._exitSync();
+    }
+  }
+
+  exitAndReturn(callback, ...args) {
+    const resource = getCurrentResource();
+    if (!resource) {
+      return callback(...args); // Let's do nothing here, there is no context
+    }
+    const store = resource[this.kResourceStore];
+    delete resource[this.kResourceStore];
+    try {
+      return callback(...args);
+    } finally {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  /**
+   * get the current store, i.e. the unique map associated with current
+   * async resource and (maybe) inherited from the parent async resource
+   * will return null if no map is currently associated
+   */
+  getStore() {
+    const resource = getCurrentResource();
+    if (resource) {
+      return resource[this.kResourceStore];
+    }
+  }
+
+  /**
+   * after this point calling getStore will return a map
+   * if called within the callback chain of another start call a new map will
+   * replace the previous one.
+   */
+  run(callback, ...args) {
+    this._enterSync();
+    process.nextTick(() => callback(...args));
```

Comment:
Same story as in https://github.com/nodejs/node/pull/26540/files#r375079149, but even worse as the user doesn't have to call `process.nextTick()` in the callback.